docs/source/customizations/native.mdx

```diff
@@ -5,7 +5,183 @@ description: Building and compiling a custom plugin for the Apollo Router with R
+
+#### Router (router_service)
+
+Implement this function if your plugin must interact at the earliest/latest stage of the pipeline. For example, this would be a good place to perform tasks such as JWT verification before allowing a Request to proceed further.
+
+#### Query Planning (query_planning_service)
+
+Implement this function if your plugin must interact with the query planning functionality. For example,  ???
+
+#### Execution (execution_service)
+
+Implement this function if your plugin must interact with the execution functionality. For example,  ???
+
+#### Subgraph (subgraph_service)
+
+Implement this function if your plugin must interact with the subgraph functionality. For example,  you wish to insert additional headers to pass to a subgraph.
+
+Typical use cases will involve interactions with the router or subgraph services. 
+
+#### Implementing your plugin
+We are going to use the hello world example plugin: examples/hello-world to demonstrate our approach.
+
+##### Plugin Source
+
+It's expected that plugins will be located in the apollo-router/src/plugins directory. Add your plugin source here and update the mod.rs to include your plugin.
+
+##### Registering your plugin
+
+The apollo-router-core crate provides the register_plugin!() macro which registers your plugin at compile time. This takes 3 arguments:
+ - a group name
+ - a name
+ - a struct implementing the Plugin trait
+
+For example:
+```
+register_plugin!("example", "hello_world", HelloWorld);
+```
+
+Choose a group name that represents your organisation and a name that represents the functionality of your plugin.
+
+###### Plugin Configuration
+
+Once you have registered your plugin, you can add configuration for it to the router configuration with an entry in the `plugins:` section. For example:
+```
+plugins:
+  example.hello_world:
+  # Any additional details here are passed to the plugin as part of your configuration
+```
+
+##### Using modules
+
+Minimally, you are going to want to use the Plugin trait, register_plugin macro and the various structs which represent Requests/Responses at different stages of processing. In addition, the Router relies heavily on standard technologies to provide useful capabilities to your plugin, for instance serde (for serializing/de-serializing data), tower (for manipulating services). Typically you will want to use at least this set of modules:
+
+```
+use apollo_router_core::plugin::Plugin;
+use apollo_router_core::{
+    register_plugin, ExecutionRequest, ExecutionResponse, QueryPlannerRequest,
+    QueryPlannerResponse, RouterRequest, RouterResponse, SubgraphRequest, SubgraphResponse,
+};
+use schemars::JsonSchema;
+use serde::Deserialize;
+use tower::util::BoxService;
+use tower::{BoxError, ServiceBuilder, ServiceExt};
+```
```

Comment:
I really like it! Maybe we could add a "Cargo.toml" starter kit like "use serde with the derive feature, tokio with the full features, schemars etc etc" ?

Reply:
It's a good suggestion, but plugin development is usually in the context of Cargo.toml for apollo-router, so I guess that is already set.

Reply:
true, stgm